l1-contracts/test/validator-selection/ValidatorSelectionBase.sol

```diff
@@ -0,0 +1,134 @@
+
+import {TimeCheater} from "../staking/TimeCheater.sol";
+// solhint-disable comprehensive-interface
+
+/**
+ * We are using the same blocks as from Rollup.t.sol.
+ * The tests in this file is testing the sequencer selection
+ */
+contract ValidatorSelectionTestBase is DecoderBase {
+  using MessageHashUtils for bytes32;
+  using EpochLib for Epoch;
+
+  struct StructToAvoidDeepStacks {
+    uint256 needed;
+    address proposer;
+    bool shouldRevert;
+  }
+
+  SlashFactory internal slashFactory;
+  Slasher internal slasher;
+  Inbox internal inbox;
+  Outbox internal outbox;
+  Rollup internal rollup;
+  MerkleTestUtil internal merkleTestUtil;
+  TestERC20 internal testERC20;
+  RewardDistributor internal rewardDistributor;
+  Signature internal emptySignature;
+  TimeCheater internal timeCheater;
+  mapping(address attester => uint256 privateKey) internal attesterPrivateKeys;
+  mapping(address proposer => uint256 privateKey) internal proposerPrivateKeys;
+  mapping(address proposer => address attester) internal proposerToAttester;
+  mapping(address => bool) internal _seenValidators;
+  mapping(address => bool) internal _seenCommittee;
+
+  /**
+   * @notice  Set up the contracts needed for the tests with time aligned to the provided block name
+   */
+  modifier setup(uint256 _validatorCount) {
+    string memory _name = "mixed_block_1";
+    {
+      DecoderBase.Full memory full = load(_name);
+      uint256 slotNumber = full.block.decodedHeader.slotNumber;
+      uint256 initialTime =
+        full.block.decodedHeader.timestamp - slotNumber * TestConstants.AZTEC_SLOT_DURATION;
+
+      timeCheater = new TimeCheater(
+        address(rollup),
+        initialTime,
+        TestConstants.AZTEC_SLOT_DURATION,
+        TestConstants.AZTEC_EPOCH_DURATION
+      );
+      vm.warp(initialTime);
+    }
+
+    CheatDepositArgs[] memory initialValidators = new CheatDepositArgs[](_validatorCount);
+
+    for (uint256 i = 1; i < _validatorCount + 1; i++) {
+      initialValidators[i - 1] = createDepositArgs(i);
+    }
+
+    testERC20 = new TestERC20("test", "TEST", address(this));
+    Registry registry = new Registry(address(this), testERC20);
+    rewardDistributor = RewardDistributor(address(registry.getRewardDistributor()));
+    rollup = new Rollup({
+      _feeAsset: testERC20,
+      _rewardDistributor: rewardDistributor,
+      _stakingAsset: testERC20,
+      _governance: address(this),
+      _genesisState: TestConstants.getGenesisState(),
+      _config: TestConstants.getRollupConfigInput()
+    });
+    slasher = Slasher(rollup.getSlasher());
+    slashFactory = new SlashFactory(IValidatorSelection(address(rollup)));
+
+    testERC20.mint(address(this), TestConstants.AZTEC_MINIMUM_STAKE * _validatorCount);
+    testERC20.approve(address(rollup), TestConstants.AZTEC_MINIMUM_STAKE * _validatorCount);
+
+    if (_validatorCount > 0) {
+      rollup.cheat__InitialiseValidatorSet(initialValidators);
```

Comment:
I think we could get rid of the cheat after this pr, but should be done fully separately, no need to grow this one as there is a bunch of node and test stuff that would change